collections/src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,519 @@
+    copySliceToArray(srcStart = 0, dest = dest, destStart = destStart, maxItems = len)
+
+  override def className = "ArrayDeque"
+
+  override def toArray[B >: A: ClassTag] =
+    copySliceToArray(srcStart = 0, dest = new Array[B](length), destStart = 0, maxItems = length)
+
+  /**
+    * This is a more general version of copyToArray - this also accepts a srcStart unlike copyToArray
+    * This copies maxItems elements from this collections srcStart to dest's destStart
+    * If we reach the end of either collections before we could copy maxItems, we simply stop copying
+    *
+    * @param dest
+    * @param srcStart
+    * @param destStart
+    * @param maxItems
+    */
+  def copySliceToArray(srcStart: Int, dest: Array[_], destStart: Int, maxItems: Int): dest.type = {
+    requireBounds(destStart, 0, dest.length)
+    val toCopy = Math.min(maxItems, Math.min(length - srcStart, dest.length - destStart))
+    if (toCopy > 0) {
+      requireBounds(srcStart)
+      val startIdx = start_+(srcStart)
+      val block1 = Math.min(toCopy, array.length - startIdx)
+      Array.copy(src = array, srcPos = startIdx, dest = dest, destPos = destStart, length = block1)
+      val block2 = toCopy - block1
+      if (block2 > 0) Array.copy(src = array, srcPos = 0, dest = dest, destPos = destStart + block1, length = block2)
+    }
+    dest
+  }
+
+  /**
+    * Trims the capacity of this ArrayDeque's instance to be the current size
+    */
+  def trimToSize(): Unit = resize(length - 1)
+
+  // Utils for common modular arithmetic:
+  @inline private[this] def start_+(idx: Int) = (start + idx) & (array.length - 1)
+  @inline private[this] def start_-(idx: Int) = (start - idx) & (array.length - 1)
+  @inline private[this] def end_+(idx: Int) = (end + idx) & (array.length - 1)
+  @inline private[this] def end_-(idx: Int) = (end - idx) & (array.length - 1)
+
+  @inline private[this] def isResizeNecessary(len: Int) = {
+    // Either resize if we need more cells OR we need to downsize BUT not a good idea to repeatedly resize small arrays
+    len >= (array.length - 1) || (2*len < array.length && array.length >= ArrayDeque.StableSize)
+  }
+
+  @inline private[this] def _get(idx: Int): A = array(start_+(idx)).asInstanceOf[A]
+
+  @inline private[this] def _set(idx: Int, elem: A) = array(start_+(idx)) = elem.asInstanceOf[AnyRef]
+
+  private[this] def resize(len: Int) = if (isResizeNecessary(len)) {
+    val n = length
+    val array2 = copySliceToArray(srcStart = 0, dest = ArrayDeque.alloc(len), destStart = 0, maxItems = n)
+    reset(array = array2, start = 0, end = n)
+  }
+
+  @inline private[this] def requireBounds(idx: Int, from: Int = 0, until: Int = length) =
+    if (idx < 0 || until <= idx) throw new IndexOutOfBoundsException(idx.toString)
+}
+
+class ArrayDequeView[A](val parent: ArrayDeque[A]) extends IndexedView[A] {
+
+  override def apply(idx: Int) = parent(idx)
+
+  override def className = "ArrayDequeView"
+}
```

Comment:
It seems to be exactly the same as the default `IndexedView` instance for `IndexedSeq` (except for the `className`).